cli_tools/import_precheck/check_osversion.go

```diff
@@ -15,43 +15,89 @@ package main
 import (
 	"fmt"
 	"strings"
+
+	"github.com/GoogleCloudPlatform/osconfig/osinfo"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/distro"
+	daisy_utils "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/daisy"
 )
 
-var osVersions = map[string][]string{
-	"centos":  {"6", "7"},
-	"debian":  {"8", "9"},
-	"ol":      {"6", "7"},
-	"rhel":    {"6", "7"},
-	"ubuntu":  {"14.04", "16.04"},
-	"windows": {"6.1", "6.3", "10.0"},
+type osVersionCheck struct {
+	osInfo *osinfo.OSInfo
 }
 
-type osVersionCheck struct{}
-
 func (c *osVersionCheck) getName() string {
 	return "OS Version Check"
 }
 
 func (c *osVersionCheck) run() (*report, error) {
 	r := &report{name: c.getName()}
-	var versions []string
-	var ok bool
-	r.Info(fmt.Sprintf("OSInfo: %+v", osInfo))
-
-	if versions, ok = osVersions[osInfo.ShortName]; !ok {
-		r.Fatal(osInfo.ShortName)
+	// Find osID from OS config's detection results.
+	major, minor := splitOSVersion(c.osInfo.Version)
+	osID, err := c.getOSID(major, minor, r)
+	if err != nil {
+		r.Info("Cannot determine OS. For supported versions, " +
+			"see https://cloud.google.com/sdk/gcloud/reference/compute/images/import")
+		r.skipped = true
 		return r, nil
 	}
 
-	found := false
-	for _, version := range versions {
-		if strings.Contains(osInfo.Version, version) {
-			found = true
+	// Check whether the osID is supported for import.
+	// Some systems are only available as BYOL, so check for both osID variants.
+	var supported bool
+	for _, suffix := range []string{"", "-byol"} {
+		if daisy_utils.ValidateOS(osID+suffix) == nil {
+			supported = true
 			break
 		}
 	}
-	if !found {
-		r.Fatal(fmt.Sprintf("version: %q not supported", osInfo.Version))
+	if supported {
+		if c.osInfo.ShortName == osinfo.Windows {
+			// Emit the NT version for Windows, since the same NT version is
+			// either Desktop or Server, and we don't want to emit a misleading message.
+			r.Info(fmt.Sprintf("Detected Windows version number: NT %s", c.osInfo.Version))
+		} else {
+			r.Info(fmt.Sprintf("Detected system: %s", osID))
+		}
+	} else {
+		r.Fatal(createFailureMessage(osID).Error())
 	}
 	return r, nil
 }
+
+func (c *osVersionCheck) getOSID(major string, minor string, r *report) (osID string, err error) {
+	if c.osInfo.ShortName == osinfo.Windows {
+		maj, min, err := distro.WindowsServerVersionforNTVersion(major, minor)
+		if err == nil {
+			major, minor = maj, min
+		}
+	}
+	release, err := distro.FromComponents(c.osInfo.ShortName, major, minor, c.osInfo.Architecture)
+	if err != nil {
+		return "", err
+	}
+	if release == nil {
+		return "", createFailureMessage("Your OS")
```

Comment:
Can we print the actual OS info, even if not supported, instead of "Your OS"?

Reply:
done!